### DISCUSSION TITLE
Setting a property with property changed signal from inherited server class

### DISCUSSION THREAD

**Discussion by colazzo** [python-sdbus/python-sdbus#47] [General]
Hello,

Is there a way to set a read-only property from the server class that implements the interface without changing the generated interface file?

Such property is typically set at runtime by the server that implements the interface, and it is not enough to use a member variable directly because in most cases you want the 'property changed' signal being fired when the property is set internally by the server.

I guess the `@sdbus.setter_private` can only be defined in the interface class in order to be able to use `set_async()` on the property?
But then you would need to manually change the interface file after having generated from xml.

Is there a way to set property (with property changed signal) without changing the generated file?

Ideally you can define everything in the xml, run the generator and implement the interface for your dbus service.
An idea would be to generate the `@sdbus.setter_private` in case the property is defined with `access="read"`, 
and provide possibility to override the private setter.
Same for `access="readwrite"` properties where you could have the `@sdbus.setter` already generated.

**Comment by igo95862** [comment]
Hello @colazzo 

Have you tried using the inheritance and `@dbus_property_async_override`?

See this unit test:

https://github.com/python-sdbus/python-sdbus/blob/9040f153d9cb88ad066e52af712792419086d634/test/test_sd_bus_async.py#L315-L327

I have not tested it but it should be possible to use the overridden property and the `setter_private` to implement this.

**Reply by igo95862** [reply]
Never mind. Looks like it is not possible to use the private setter from overload:

```
AttributeError: 'DbusOverload' object has no attribute 'setter_private'
```

However, this can probably be fixed.

**Reply by igo95862** [reply]
Meanwhile the options are using the regular `setter` or emitting the properties changed manually.

**Reply by cleersnijder** [reply]
Hi igo95862, thank you having investigated this! 

What is the easiest way to emit in async way the property changed from the derived class? Can you give an example?

**Reply by igo95862** [reply]
Hello @cleersnijder 

Have you tried the `dbus_property_async_override` ?

See this unit test: https://github.com/python-sdbus/python-sdbus/blob/9040f153d9cb88ad066e52af712792419086d634/test/test_sd_bus_async.py#L310

**Reply by colazzo** [reply]
Hi @igo95862,

Tested your proposed solution: it changes the access of the property from read-only to read/write, so this is not an option.

It seems like the private setter can only be defined in the interface file, but I would like to avoid that. 
I don't mind sending the changed signal from the subclass manually, but how can this be done without having to add something in the interface file?

**Reply by igo95862** [reply]
I should fix `setter_private` not being available in overridden properties.

> I don't mind sending the changed signal from the subclass manually

https://github.com/python-sdbus/python-sdbus/issues/33#issuecomment-1374482952

**Reply by colazzo** [reply]
That would be great, thank you.

**Reply by colazzo** [reply]
I actually use now a work-around by using a private method inside the sub-class which first sets the new property value, followed by the call to setter_private in the parent class using super(). I just need to first add manually the corresponding @<property>.setter_private methods in the generated class file.

For future version it would be nice to have these private methods generated for the properties with access="read",
and for the ones with "readwrite" you would get the corresponding @sdbus.setter methods.

**Reply by igo95862** [reply]
@colazzo I added a property override private setter in 513fd920f496939dd72cb318f07003c6216995c5. Could you give it a test?

I also made a significant rework of internal meta classes. The code existing code using public interfaces should work exactly the same but some implementation details have changed.

**Reply by colazzo** [reply]
Hi @igo95862 I did a test locally and it seems to be working fine: The Property stays read-only and I can set new value using the `set_async` method from derived class which triggers the property changed signal. 
Thanks for that!
